File: src/main/java/ac/quant/quickfixspec/inspections/ReplaceWithDefinitionInspection.java
package ac.quant.quickfixspec.inspections;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.codeInspection.ProblemDescriptor;
import com.intellij.codeInspection.LocalQuickFix;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.SmartPointerManager;
import com.intellij.psi.SmartPsiElementPointer;
import com.intellij.psi.xml.XmlTag;
import com.intellij.util.IncorrectOperationException;
import org.jetbrains.annotations.NotNull;

import static ac.quant.quickfixspec.common.PsiUtils.*;

public class ReplaceWithDefinitionInspection extends LocalInspectionTool {

    @Override
    public @NotNull PsiElementVisitor buildVisitor(@NotNull ProblemsHolder holder, boolean isOnTheFly) {
        return new PsiElementVisitor() {
            @Override
            public void visitElement(@NotNull PsiElement element) {
                if (isNotComponentTagReference(element)) {
                    return;
                }

                final XmlTag rootTag = getRootTag(element);
                final XmlTag tag = (XmlTag) element;
                if (rootTag == null) {
                    return;
                }

                final String componentName = tag.getAttributeValue("name");


                XmlTag definition = findDefinition(componentName, tag.getName(), rootTag);
                if (definition != null) {
                    holder.registerProblem(tag, "Replace with definition", new ReplaceWithDefinitionQuickFix(definition));
                }
            }
        };
    }

    private boolean isNotComponentTagReference(@NotNull PsiElement element) {
        if (!(element instanceof XmlTag tag)) {
            return true;
        }

        final XmlTag parentTag = tag.getParentTag();
        if (parentTag == null) {
            return true;
        }

        if (parentTag.getName().equals("components")) {
            return true;
        }

        // if the tag is not self-closing that means it is a definition
        if (tag.getSubTags().length > 0) {
            return true;
        }

        return !tag.getName().equals("component");
    }

    private static class ReplaceWithDefinitionQuickFix implements LocalQuickFix {
        @SafeFieldForPreview
        private final SmartPsiElementPointer<XmlTag> definitionPointer;

        ReplaceWithDefinitionQuickFix(XmlTag definition) {
            this.definitionPointer = SmartPointerManager.createPointer(definition);
        }

        @Override
        public @NotNull String getName() {
            return "Replace with definition";
        }

        @Override
        public @NotNull String getFamilyName() {
            return getName();
        }

        @Override
        public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) throws IncorrectOperationException {
            // Get the <component> tag that is to be replaced
            XmlTag tagToReplace = (XmlTag) descriptor.getPsiElement();

            // Retrieve the full definition of the <component> from the pointer
            XmlTag definition = definitionPointer.getElement();

            if (definition != null) {
                // Create a copy of the definition tag to be inserted
                XmlTag newTag = (XmlTag) definition.copy();

                // Replace the original <component> tag with the full definition
                tagToReplace.replace(newTag);
            }
        }
    }
}

File: src/main/kotlin/ac/quant/quickfixspec/inspections/ReplaceWithDefinitionInspection.kt
package ac.quant.quickfixspec.inspections

import ac.quant.quickfixspec.common.spec.XmlUtils.findDefinition
import ac.quant.quickfixspec.common.spec.XmlUtils.getRootTag
import ac.quant.quickfixspec.common.spec.XmlUtils.isComponentReference
import ac.quant.quickfixspec.common.spec.XmlUtils.getCurrentTag

import ac.quant.quickfixspec.inspections.ReplaceWithDefinitionInspection.ReplaceWithDefinitionQuickFix
import com.intellij.codeInsight.intention.FileModifier.SafeFieldForPreview
import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.LocalQuickFix
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.SmartPointerManager
import com.intellij.psi.SmartPsiElementPointer
import com.intellij.psi.xml.XmlTag
import com.intellij.util.IncorrectOperationException


class ReplaceWithDefinitionInspection : LocalInspectionTool() {
    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
        return object : PsiElementVisitor() {
            override fun visitElement(element: PsiElement) {
                val rootTag = getRootTag(element) ?: return

                val currentTag: XmlTag = getCurrentTag(element)
                if (currentTag.name != "component") {
                    return
                }


                // return if the tag is not a component reference, and it's actually a definition from the "components" group
                if (!isComponentReference(currentTag)) {
                    return
                }

                val componentName = currentTag.getAttributeValue("name") ?: return
                val definition = findDefinition(componentName, "component",  rootTag) ?: return

                // check if we didn't already register a problem for this tag
                if (holder.isOnTheFly && holder.results.any { it.psiElement == currentTag }) {
                    return
                }

                holder.registerProblem(currentTag, "Replace with definition", ReplaceWithDefinitionQuickFix(definition))
            }
        }
    }

    private class ReplaceWithDefinitionQuickFix(definition: XmlTag) : LocalQuickFix {
        @SafeFieldForPreview
        private val definitionPointer: SmartPsiElementPointer<XmlTag?> = SmartPointerManager.createPointer<XmlTag?>(definition)

        override fun getName(): String {
            return "Replace component reference with definition"
        }

        override fun getFamilyName(): String {
            return name
        }

        @Throws(IncorrectOperationException::class)
        override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
            // Get the <component> tag that is to be replaced
            val tagToReplace = descriptor.psiElement as XmlTag

            // Retrieve the full definition of the <component> from the pointer
            val definition = definitionPointer.getElement()

            if (definition != null) {
                // Create a copy of the definition tag to be inserted
                val newTag = definition.copy() as XmlTag

                // Replace the original <component> tag with the full definition
                tagToReplace.replace(newTag)
            }
        }
    }
}